uro98/assignment2/PrintAncestors.java

```diff
@@ -0,0 +1,84 @@
+import java.util.Arrays;
+import java.util.LinkedList;
+
+public class PrintAncestors {
+  
+  /**
+   * Print the ancestors of the node with the given key in a binary tree as a list.
+   * Uses left-to-right depth first search.
+   * An empty list is printed if the node is not found.
+   * @param tree  a generic binary tree
+   * @param key   the target value of a node
+   */
+  public static <T> void printAncestors(BinaryTree<T> tree, T key) {
+    LinkedList<T> ancestors = new LinkedList<T>();
+    LinkedList<Node<T>> frontier = new LinkedList<Node<T>>();
+    
+    frontier.addFirst(tree.root);
+    
+    while (!frontier.isEmpty()) {
+      Node<T> node = frontier.pop();
+      
+      if (node.data.equals(key)) {
+        printList(ancestors);
+        return;
+      } else {
+        ancestors.addFirst(node.data);
+        
+        // Push the children of the node to the frontier stack, push the right child first to make DFS search left to right
+        if (node.right != null) {
+          frontier.addFirst(node.right);
+        }
+        if (node.left != null) {
+          frontier.addFirst(node.left);
+        }
+        
+        /* If the next node to be explored in the frontier is further up the tree, remove the nodes that aren't ancestors
+         * (ie the ones on the same level and below)
+         */
+        if (!frontier.isEmpty() && node.depth >= frontier.peek().depth) {
+          int depthDifference = node.depth - frontier.peek().depth;
+          for (int i = 0; i < depthDifference + 1; i++) {
+            ancestors.pop();
+          }
+        }
+      }
+    }
+    
+    // If the node is not found, print an empty list
+    ancestors.clear();
+    printList(ancestors);
```

Comment:
Optional: instead of sequence `ancestors.clear(); printList(ancestors);` you could just skip clearing and do `printList(Collections.emptyList())`, which would make it explicit what is happening.